### PR TITLE
Strings API escape sequences

### PR DESCRIPTION
**Pull request by ajreynol**
This changes constructors of CVC4::String to not convert escape sequences by default.

We convert escape sequences in strings now only for strings that come from the parsers.

This addresses #1169.

### CODE REVIEW

**Comment by ajreynol** [comment]
Kshitij has not yet confirmed this fix works for him.

Looking into this further, it appears that the previous behavior was nonsensical.  For example, when a string is copied, we could actually re-remove escape sequences.  So that something like "\\n" would initially be converted at parse time to "\n" which upon copying (for instance in a rewriter), this would be converted to a newline.

**Comment by ajreynol** [comment]
Virtually no effect on the CMU CAV17 set: https://www.starexec.org/starexec/secure/details/job.jsp?id=24564 (CVC4-101317-4-stringsEscSeq)

**Comment by kbansal** [comment]
> Although looking into this further, it appears that the previous behavior was nonsensical. For example, when a string is copied, we could actually re-remove escape sequences. So that something like "\\n" would initially be converted at parse time to "\n" which upon copying (for instance in a rewriter), this would be converted to a newline.

Good catch. Does the commit fix this yet? (didn't look like it -- since you perhaps just noticed after commit -- but wanted to confirm before digging into rest of the changes)

Two other general comments:
1. Is there a method to convert back? (effectively run convertUnsignedIntToChar over internal rep)
2. Should the logic for escaping/unescaping lie in the parser? Or, if we want to keep it this class, maybe have a separate method for handling the escaping? (e.g. CVC4::Strings::UnescapeAndBuild(...) or CVC4::Strings::BuildFromEscapedString(...) or something like that?) Would address your observation I think.

File: src/util/regexp.h [hunk 1]
```diff
@@ -45,8 +45,10 @@ class CVC4_PUBLIC String {
   }
 
   String() = default;
-  explicit String(const std::string& s) : d_str(toInternal(s)) {}
-  explicit String(const char* s) : d_str(toInternal(std::string(s))) {}
+  explicit String(const std::string& s, bool useEscSequences = false) :
```
Comment (timothy-king) [inline src/util/regexp.h]: Let's document what useEscSequences does.

File: src/util/regexp.h [hunk 2]
```diff
@@ -45,8 +45,10 @@ class CVC4_PUBLIC String {
...
+   d_str(toInternal(s, useEscSequences)) {}
+  explicit String(const char* s, bool useEscSequences = false) :
```
Comment (timothy-king) [inline src/util/regexp.h]: nit: Pretty sure the ":" will end up on the other line post clang-format.

**Comment by ajreynol** [comment]
The commit does fix the problem I mention.  I realized this only after I committed :)

Regarding 1, there is a "toString()" function in CVC4::String.  It looks like this had the same behavior of automatically processing escape sequences.  My latest commit has made this symmetric to the constructor. Hence, we take a useEscSequences, by default false.

Regarding 2, the issue I brought up is fixed, but I'm not sure yet if we should move the code to the printer.

**Pull request merged**